src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventSink.cs

```diff
@@ -0,0 +1,243 @@
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnBeforeOpenDocument;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnInitializedDocument;
+        public event EventHandler<RunningDocumentTableRenamedEventArgs> OnRenameDocument;
+
+        [ImportingConstructor]
+        public RunningDocumentTableEventSink(IThreadingContext threadingContext, IVsEditorAdaptersFactoryService editorAdaptersFactoryService, SVsServiceProvider serviceProvider)
+        {
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: true);
+            _runningDocumentTable = (IVsRunningDocumentTable4)serviceProvider.GetService(typeof(SVsRunningDocumentTable));
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            ((IVsRunningDocumentTable)_runningDocumentTable).AdviseRunningDocTableEvents(this, out _runningDocumentTableEventsCookie);
+        }
+
+        public int OnAfterFirstDocumentLock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnBeforeLastDocumentUnlock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+        {
+            if (dwReadLocksRemaining + dwEditLocksRemaining == 0)
+            {
+                if (CheckPreconditions(docCookie))
+                {
+                    var args = new RunningDocumentTableEventArgs(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie));
+                    OnCloseDocument?.Invoke(this, args);
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterSave(uint docCookie)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnAfterAttributeChange(uint docCookie, uint grfAttribs)
+        {
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterAttributeChangeEx(uint docCookie, uint grfAttribs, IVsHierarchy pHierOld, uint itemidOld, string pszMkDocumentOld, IVsHierarchy pHierNew, uint itemidNew, string pszMkDocumentNew)
+        {
+            // Did we rename?
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_MkDocument) != 0)
+            {
+                if (CheckPreconditions(docCookie))
+                {
+                    var args = new RunningDocumentTableRenamedEventArgs(docCookie, pszMkDocumentNew, pszMkDocumentOld);
+                    OnRenameDocument?.Invoke(this, args);
+                }
+            }
+
+            if ((grfAttribs & (uint)__VSRDTATTRIB3.RDTA_DocumentInitialized) != 0)
+            {
+                if (TryGetBuffer(docCookie, out var buffer))
+                {
+                    var args = new RunningDocumentTableInitializedEventArgs(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie), buffer);
```

Comment:
so, why some we check moniker being null and not others? if we check moniker being null and bail out then, is there any point this even raise event when moniker is null?

Reply:
Was trying to keep the same existing behavior as much as possible, but I can go back and make the checks consistent.

Reply:
@jasonmalinowski is it expected in some cases that the moniker is null and it's OK to proceed in the misc workspace?  It looks like whenever the document is going through TryUntrackClosingDocument the moniker isn't necessarily used (so this ends up being rename and close).  If I did a null check before even sending the event to the misc workspace, would that be bad?

The VS one seems like it should never be null, as it will just return an empty documentIds list if the moniker isn't present.

Reply:
I cannot think of a specific reason why these can't be made consistent. Honestly I'm not sure in what scenario the moniker would be null, but whatever it would be, it's something that we definitely don't want other things touching.